kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -156,6 +165,33 @@ public Snapshot getSnapshotForTimestamp(
   // Helper Methods //
   ////////////////////
 
+  /**
+   * Creates a metrics-enabled engine that wraps the provided engine to collect filesystem listing
+   * call metrics.
+   */
+  private Engine createMetricsEngine(Engine engine, SnapshotQueryContext snapshotContext) {
+    return new DelegateEngine(engine) {
+      @Override
+      public FileSystemClient getFileSystemClient() {
+        return new DelegateFileSystemClient(super.getFileSystemClient()) {
+          @Override
+          public CloseableIterator<FileStatus> listFrom(String filePath) throws IOException {
+            long startTime = System.nanoTime();
+            try {
+              return super.listFrom(filePath);
```

Comment:
does all contents of the list come in one call? afaik, there is a pagination involved in the list iterator.

Reply:
Discussed offline, we should not make assumption that `listFrom` returns all the data. Given it returns an iterator, there could be a case that actual listing happens partially or fully and lazily during iterating iterator. I updated this PR only collect counter